### ISSUE TITLE
Nix gives errors due to paths without separators on package installation

### ISSUE THREAD

**Issue by denbeigh2000** [NixOS/nix#6345]
**Describe the bug**

When I attempt to upgrade certain packages, I observe failures like this:
```
error: getting status of '/nix/store/lgk3b7p9b8qgdvrqglym0icay68789lg-zulu11.48.21-ca-jdk-11.0.11/zulu-11.jdk/Contents/Home/includedarwin': No such file or directory
error: builder for '/nix/store/6vkd1xs7v3zbj540q5dv040j7x4j8by1-user-environment.drv' failed with exit code 1
```
(`nixpkgs-1882c6b7368fd284ad01b0a5b5601ef136321292 jdk11`)

```
building '/nix/store/7n876dfzayr4nvxb6nd1rp9cy10vpd1m-user-environment.drv'...
error: getting status of '/nix/store/07aza66d3s6as1nah4l36c28hxn7gm6l-pkg-config-0.29.2/shareaclocal': No such file or directory
error: builder for '/nix/store/7n876dfzayr4nvxb6nd1rp9cy10vpd1m-user-environment.drv' failed with exit code 1;
```

(`nixpkgs-1882c6b7368fd284ad01b0a5b5601ef136321292 pkg-config`)

In both cases, the paths `nix` intended to access were missing their final path separators, correct paths were `include/darwin` and `share/aclocal`.

**Steps To Reproduce**

I don't a minimal case to reproduce this, as this was extracted from a large, integrated setup.

The steps we see this under:
 - Install a package set.
 - Upgrade version of `nixpkgs` (in our case, we upgraded from `21.05pre282880.5268ee2ebac` to `1882c6b7368fd284ad01b0a5b5601ef136321292`)
 - Attempt to install the same package set.
 - The packages are not installed successfully, and show above errors with malformed paths.

**Expected behavior**

I expect these packages to be installed without any invalid path errors.

**`nix-env --version` output**

```
$ nix-env --version
nix-env (Nix) 2.6.0
```

**Additional context**

If I use the `--remove-all` flag, I don't see this occur.
If I start from a fresh state (either on a new machine, or by wiping my nix installation), I don't see this occur.

**Comment by pennae** [comment]
could be a duplicate of #6017 (should be fixed 2.6.1)

**Comment by denbeigh2000** [comment]
> could be a duplicate of #6017 (should be fixed 2.6.1)

Looks like it is this issue, thanks for the tip!
